### DISCUSSION TITLE
Does G+Smo support or plans to support T-Splines?

### DISCUSSION THREAD

**Discussion by andre-caldas** [gismo/gismo#737] [Q&A]
I saw G+Smo supports some different types of splines. But I do not really understand their differences. Sorry if I am asking something silly!

Does G+Smo support or plans to support T-Splines?
The patent, by the way, has expired in March.
https://en.wikipedia.org/wiki/T-spline

**Comment by andre-caldas** [comment]
Sorry for the silly question above! I realize that G+Smo people are more inclined to the THB-spline thing.

I will start reading more about THB-spline. I am planing to write a CAD software, and I am still tying to understand all the alternatives so I can pick one up. :-)

Does anyone of you guys here have any reading / studying recommendations?

I have just found Angelos' _"THB-splines: An effective mathematical technology for adaptive refinement in geometric design and isogeometric analysis"_. I will start reading it right now.

**Comment by ckarampa** [comment]
Hello Andre, 

as far as I know there are no "official" plans to implement T-Splines in G+Smo, but it is not something irrelevant either. I believe that the way the project is structured, one can implement a variety of spline constructions and as long as a basic interface is maintained, they can take advantage of most of the rest of gismo. If you are familiar with T-Splines, or are willing to study and implement it, I think it would be a significant contribution to the library.

**Reply by andre-caldas** [reply]
I do not think I have what it takes, right now. It is all very new to me. I knew a little about Bézier curves because I had read something in a book some 30 years ago, when I was a teenager. :-)

I am planning to write this CAD software of mine, so I started to look for a "universal" data structure to use, pretty much as Rhino has adopted NURBS. Right now, I am just understanding the available options. I thought the "ideal thing" was NURBS. Then I heard about T-splines. Then manifold T-splines, then orbifolds...

I am starting to think that THB-splines offer the best compromise. The T-splines things sounds a little "hackish" to me. The results are good, but the design decisions sound a little too "trial-and-error" / empirical / heuristic. I don't think people have fully decided on how to handle the so called "exceptional points". Definitions are not clearly stated and the theory is developed in a very procedural way (extend the line until it reaches...). I am looking for something with sound rock solid mathematical base.

My liking T-splines or not, I do hope I will fully understand it through the whole process. I do hope I will understand G+Smo's internals. You can count on me, then. Anyone who is already planing / implementing it can also count on me. :-)

**Comment by hverhelst** [comment]
Hi @andre-caldas ,

Nice to see that you are interested in using G+Smo for your project! 
I have worked quite a bit with THB-splines and we are currently developing a new module where the implementation of new spline constructions (e.g. the [decoupled basis](https://doi.org/10.1016/j.cagd.2014.05.004)) can be developed quite easily. The new module is based on [this paper](https://doi.org/10.1007/978-3-319-67885-6_3). It is currently in the private part of the G+Smo repo.
If you want to know more, we could schedule a meeting together!

**Reply by andre-caldas** [reply]
Wow! That makes me very happy!

I will read those two papers before. Then, I meeting would be great. I will contact you, if you do not mind.
